internal/db/schema/migrations/postgres/2/87_iam_test.go

```diff
@@ -0,0 +1,110 @@
+package migration
+
+import (
+	"context"
+	"database/sql"
+	"testing"
+
+	"github.com/hashicorp/boundary/internal/auth/password"
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/db/schema"
+	"github.com/hashicorp/boundary/internal/docker"
+	"github.com/hashicorp/boundary/internal/iam"
+	"github.com/jinzhu/gorm"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_PrimaryAuthMethodChanges(t *testing.T) {
+	t.Parallel()
+	const priorMigration = 2086
+	const primaryAuthMethodMigration = 2087
+	t.Run("migrate-store", func(t *testing.T) {
+		// this is a sequential test which relies on:
+		// 1) initializing the db using a migration up to the "priorMigration"
+		//
+		// 2) seeding the database with some test scopes and auth methods
+		//
+		// 3) running the migrations that sets the primary auth methods for
+		// existing scopes which only have one auth method.
+		//
+		// 4) asserting some bits about the state of the db.
+		assert, require := assert.New(t), require.New(t)
+		dialect := "postgres"
+		ctx := context.Background()
+
+		c, u, _, err := docker.StartDbInDocker(dialect)
+		require.NoError(err)
+		t.Cleanup(func() {
+			require.NoError(c())
+		})
+		d, err := sql.Open(dialect, u)
+		require.NoError(err)
+
+		// migration to the prior migration (before the one we want to test)
+		oState := schema.TestCloneMigrationStates(t)
+		nState := schema.TestCreatePartialMigrationState(oState["postgres"], priorMigration)
+		oState["postgres"] = nState
+
+		m, err := schema.NewManager(ctx, dialect, d, schema.WithMigrationStates(oState))
+		require.NoError(err)
+
+		assert.NoError(m.RollForward(ctx))
+		state, err := m.CurrentState(ctx)
+		require.NoError(err)
+		assert.Equal(priorMigration, state.DatabaseSchemaVersion)
+		assert.False(state.Dirty)
+
+		// okay, now we can seed the database with test data
```

Comment:
Isn't your assertion true for all tests?  I'm not seeing how this test is any different that all the other unit tests that use test functions from other packages.

Reply:
Just to document our conversation from off thread, The best long term fix for this type of concern is to have a better solution for integration tests which would allow us to test version to version migration steps.  Regarding this specific situation having it as a unit test like this makes sense for now since the alternative would be mutating the sql directly which would not necessarily put the data in a state that accurately represents what the code would/could put it in.

Thanks for taking the time to discuss this with me!